### PR TITLE
Vectorized aggregation with grouping by one fixed-size column

### PR DESCRIPTION
**Pull request by akuzm**
The implementation uses the Postgres simplehash hash table for by-value fixed-size compressed columns.

The biggest improvement on a "sensible" query is about 90%, and a couple of queries show bigger improvements but these are very synthetic cases that don't make much sense: 
https://grafana.ops.savannah-dev.timescale.com/d/fasYic_4z/compare-akuzm?orgId=1&var-branch=All&var-run1=3815&var-run2=3816&var-threshold=0.02&var-use_historical_thresholds=true&var-threshold_expression=2%20%2A%20percentile_cont%280.90%29&var-exact_suite_version=false&from=now-2d&to=now

### CODE REVIEW

**Review by svenklemm — APPROVED** [review]

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c [hunk 1]
```diff
@@ -0,0 +1,735 @@
...
+	 * i.e. if they store strings or float8 datum on 32-bit systems. Valid until
+	 * the grouping policy is reset.
+	 */
+	MemoryContext agg_extra_mctx;
+
+	/*
+	 * Temporary storage of aggregate state offsets for a given batch. We keep
+	 * it in the policy because it is potentially too big to keep on stack, and
+	 * we don't want to reallocate it each batch.
+	 */
```
Comment (erimatnor) [inline tsl/src/nodes/vector_agg/grouping_policy_hash.c:127]: Took me a while to figure out what these offsets are for. Could we do a better description here?

For example (if I got things right):

The agg_state_offsets array  contains an offset for each "row" in an ArrowArray that points to the aggregate states for that row. The agg state is points to is decided by the value (key) in the grouping column. Thus, the number of agg states is equal to the number of unique values in the grouping column's ArrowArray.

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c [hunk 2]
```diff
@@ -0,0 +1,735 @@
...
+
+/*
+ * For the hash table, use the generic Datum key that is mapped to the aggregate
+ * state index.
+ */
+typedef struct
+{
+	Datum key;
+	uint32 status;
+	uint32 agg_state_index;
```
Comment (erimatnor) [inline tsl/src/nodes/vector_agg/grouping_policy_hash.c:69]: What about having the Agg state in the hash table entry instead? What are the pros/cons of having it here compared to in an outside array?

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c [hunk 3]
```diff
@@ -0,0 +1,735 @@
...
+			continue;
+		}
+
+		if (key_valid)
+		{
+			bool found = false;
+			HashEntry *restrict entry = h_insert(table, key, &found);
+			if (!found)
+			{
+				entry->agg_state_index = next_unused_state_index++;
```
Comment (erimatnor) [inline tsl/src/nodes/vector_agg/grouping_policy_hash.c:293]: I am wondering why we're not initializing the aggregate states here instead of in an additional loop after the fill_offsets?

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c [hunk 4]
```diff
@@ -0,0 +1,735 @@
...
+		{
+			if (next_unused_state_index > policy->allocated_aggstate_rows)
+			{
+				policy->allocated_aggstate_rows = policy->allocated_aggstate_rows * 2 + 1;
+				forboth (aggdeflc, policy->agg_defs, aggstatelc, policy->per_agg_states)
+				{
+					VectorAggDef *agg_def = lfirst(aggdeflc);
+					lfirst(aggstatelc) =
+						repalloc(lfirst(aggstatelc),
+								 policy->allocated_aggstate_rows * agg_def->func.state_bytes);
```
Comment (erimatnor) [inline tsl/src/nodes/vector_agg/grouping_policy_hash.c:625]: Wouldn't it be nicer if the agg_def->func had an allocation function instead? Then you'd just do `agg_def->func.alloc_states(num_states)` instead.

**Review by erimatnor — COMMENTED** [review]
This is really good functionality but there's a lot of code to review and it is a bit hard to understand and follow so it takes time.

I think there are some things that can be done to improve the code and the readability so that it is easier to understand for others.

I've made some suggestions, and probably not all of them are valid or good ideas.  But at least the places I've made suggestions for so far are areas that are particularly hard to follow.

I will submit what I have so far, and will try to do the rest later.

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c [hunk 5]
```diff
@@ -0,0 +1,735 @@
...
+	 * the grouping policy is reset.
+	 */
+	MemoryContext agg_extra_mctx;
+
+	/*
+	 * Temporary storage of aggregate state offsets for a given batch. We keep
+	 * it in the policy because it is potentially too big to keep on stack, and
+	 * we don't want to reallocate it each batch.
+	 */
+	uint32 *offsets;
```
Comment (erimatnor) [inline tsl/src/nodes/vector_agg/grouping_policy_hash.c:128]: Another question: what if we made this and array of pointers to the agg states instead of an array of offsets?

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c [hunk 6]
```diff
@@ -0,0 +1,735 @@
...
+										  end_row,
+										  arg_datum,
+										  arg_isnull,
+										  agg_extra_mctx);
+		}
+		else
+		{
+			for (int i = start_row; i < end_row; i++)
+			{
+				if (offsets[i] == 0)
```
Comment (erimatnor) [inline tsl/src/nodes/vector_agg/grouping_policy_hash.c:253]: Isn't 0 a legitimate offset into the agg_states array? Or is it 1-indexed?

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c (hunk 2 above)
Comment (akuzm) [reply tsl/src/nodes/vector_agg/grouping_policy_hash.c:69]: The hash table should be as small as possible, because we're accessing its memory randomly, and the speed of this depends on whether it fits into the caches.

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c (hunk 3 above)
Comment (akuzm) [reply tsl/src/nodes/vector_agg/grouping_policy_hash.c:293]: Just following a common rule of thumb for high-throughput data processing: colocate similar items in memory, do the same operation on them in bulk.

In this grouping agorithm we have two major stages:
* Match input row to a unique key index (`offsets`). This uses the hash table and the key column data.
* Calculate the aggregate functions. This uses their input columns and their states.
These stages are done separately, and the data required for them are grouped together. According to the same principle, each aggregate function is processed for the entire batch, not all aggregate functions for one row.

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c (hunk 5 above)
Comment (akuzm) [reply tsl/src/nodes/vector_agg/grouping_policy_hash.c:128]: I want to colocate the states of one aggregate function together because they are processed together, and to use the same offsets (unique indexes of grouping key, basically) to access the states of all aggregate functions. With pointers this would be less convenient, and also they take up more memory.

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c (hunk 1 above)
Comment (akuzm) [reply tsl/src/nodes/vector_agg/grouping_policy_hash.c:127]: Yeah I'll try to describe it better. They are basically unique indexes of the grouping keys, and the array is indexed by the row index of the input batch.

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c (hunk 4 above)
Comment (akuzm) [reply tsl/src/nodes/vector_agg/grouping_policy_hash.c:625]: I think it could be slightly confusing. The aggregate functions can't decide to allocate their memory in a different way, so this is not a part of their interface. No need to make a method that has the same single-line implementation for every aggregate.

File: tsl/src/nodes/vector_agg/function/functions.c [hunk 7]
```diff
@@ -42,26 +45,44 @@ count_emit(void *agg_state, Datum *out_result, bool *out_isnull)
...
 	state->count += n;
 }
 
+static void
+count_star_many_scalar(void *restrict agg_states, uint32 *restrict offsets, int start_row,
+					   int end_row, Datum constvalue, bool constisnull,
+					   MemoryContext agg_extra_mctx)
+{
+	CountState *states = (CountState *) agg_states;
+	for (int row = start_row; row < end_row; row++)
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/function/functions.c:61]: I guess that's the C++ habit of mine where `end` is idiomatically a past-the-end invalid iterator. In general I think the ranges with exclusive right end are very common, like `[begin, end)`. Do you have a better name for this? Sometimes I write `past_the_end_row` to make it absolutely clear, but this feels a little too long for common usage...

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c [hunk 8]
```diff
@@ -0,0 +1,735 @@
...
+	 * In single-column grouping, we store the null key outside of the hash
+	 * table, and it has a reserved aggregate state index 1. We also reset this
+	 * flag after we output the null key during iteration.
+	 */
+	bool have_null_key;
+
+	/*
+	 * A memory context for aggregate functions to allocate additional data,
+	 * i.e. if they store strings or float8 datum on 32-bit systems. Valid until
+	 * the grouping policy is reset.
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/grouping_policy_hash.c:119]: I was thinking more about statistical sketches or exact number of unique values, these functions have to allocate generic variable-length data somewhere, and that's the purpose of this context.

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c [hunk 9]
```diff
@@ -0,0 +1,735 @@
...
+			DEBUG_LOG("spill after %ld input %ld valid %ld bulk filtered %.0f keys %f ratio %ld "
+					  "aggctx bytes %ld aggstate bytes",
+					  policy->stat_input_total_rows,
+					  policy->stat_input_valid_rows,
+					  policy->stat_bulk_filtered_rows,
+					  keys,
+					  policy->stat_input_valid_rows / keys,
+					  MemoryContextMemAllocated(policy->table->ctx, false),
+					  MemoryContextMemAllocated(policy->agg_extra_mctx, false));
+		}
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/grouping_policy_hash.c:692]: Yeah these are the cross-platform printf int width specifiers that I didn't want to use, as mentioned in the comments for the macro :) Although they look less ugly than I remember, are these the Postgres ones?

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c (hunk 6 above)
Comment (akuzm) [reply tsl/src/nodes/vector_agg/grouping_policy_hash.c:253]: I'm using 0 as invalid, added some comments about that.

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c [hunk 10]
```diff
@@ -0,0 +1,735 @@
...
+		 */
+		if (next_unused_state_index > last_initialized_state_index)
+		{
+			if (next_unused_state_index > policy->allocated_aggstate_rows)
+			{
+				policy->allocated_aggstate_rows = policy->allocated_aggstate_rows * 2 + 1;
+				forboth (aggdeflc, policy->agg_defs, aggstatelc, policy->per_agg_states)
+				{
+					VectorAggDef *agg_def = lfirst(aggdeflc);
+					lfirst(aggstatelc) =
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/grouping_policy_hash.c:623]: Done.

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c [hunk 11]
```diff
@@ -0,0 +1,735 @@
...
+	 */
+	uint32 *offsets;
+	uint64 num_allocated_offsets;
+
+	/*
+	 * Storage of aggregate function states, each List entry is the array of
+	 * states for the respective function from agg_defs. The state index 0 is
+	 * invalid, and the state index 1 is reserved for a null key.
+	 */
+	List *per_agg_states;
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/grouping_policy_hash.c:136]: Made it a `void**`, and `agg_defs` a proper `VectorAggDef` array.

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c [hunk 12]
```diff
@@ -0,0 +1,735 @@
...
+	Assert(!policy->returning_results);
+
+	const uint64_t *restrict filter = batch_state->vector_qual_result;
+	const int n = batch_state->total_batch_rows;
+
+	/*
+	 * Initialize the array for storing the aggregate state offsets corresponding
+	 * to a given batch row.
+	 */
+	if ((size_t) n > policy->num_allocated_offsets)
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/grouping_policy_hash.c:505]: Yes, this is done to avoid checks and allocations in the hot loop that is mapping the grouping column values to the unique indexes of grouping keys. Added a comment.

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c [hunk 13]
```diff
@@ -0,0 +1,735 @@
...
+	}
+	memset(policy->offsets, 0, n * sizeof(policy->offsets[0]));
+
+	/*
+	 * For the partial aggregation node, the grouping columns are always in the
+	 * output, so we don't have to separately look at the list of the grouping
+	 * columns.
+	 */
+	Assert(list_length(policy->output_grouping_columns) == 1);
+	GroupingColumn *g = linitial(policy->output_grouping_columns);
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/grouping_policy_hash.c:518]: Made it an array.

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c [hunk 14]
```diff
@@ -0,0 +1,735 @@
...
+		policy->offsets = palloc(sizeof(policy->offsets[0]) * policy->num_allocated_offsets);
+	}
+	memset(policy->offsets, 0, n * sizeof(policy->offsets[0]));
+
+	/*
+	 * For the partial aggregation node, the grouping columns are always in the
+	 * output, so we don't have to separately look at the list of the grouping
+	 * columns.
+	 */
+	Assert(list_length(policy->output_grouping_columns) == 1);
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/grouping_policy_hash.c:517]: I'll keep it here for now, because in the following changes the hash grouping policy will choose a different hashing strategy here based on the type of grouping key (i.e. arithmetic/text/multicolumn).

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c [hunk 15]
```diff
@@ -0,0 +1,735 @@
...
+ */
+typedef struct
+{
+	/*
+	 * We're using data inheritance from the GroupingPolicy.
+	 */
+	GroupingPolicy funcs;
+
+	List *agg_defs;
+	List *output_grouping_columns;
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/grouping_policy_hash.c:94]: Donw.

File: tsl/test/expected/vectorized_aggregation.out [hunk 16]
```diff
@@ -388,23 +388,20 @@ SELECT sum(segment_by_value) FROM testtable GROUP BY float_value;
          Output: _hyper_1_1_chunk.float_value, (PARTIAL sum(_hyper_1_1_chunk.segment_by_value))
          Workers Planned: 2
          ->  Parallel Append
-               ->  Partial HashAggregate
-                     Output: _hyper_1_1_chunk.float_value, PARTIAL sum(_hyper_1_1_chunk.segment_by_value)
-                     Group Key: _hyper_1_1_chunk.float_value
+               ->  Custom Scan (VectorAgg)
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/test/expected/vectorized_aggregation.out:391]: Done, added a "Grouping Policy" property in the verbose mode.

File: tsl/src/nodes/vector_agg/grouping_policy_hash.c [hunk 17]
```diff
@@ -0,0 +1,735 @@
...
+	const int n = batch_state->total_batch_rows;
+
+	/*
+	 * Initialize the array for storing the aggregate state offsets corresponding
+	 * to a given batch row.
+	 */
+	if ((size_t) n > policy->num_allocated_offsets)
+	{
+		policy->num_allocated_offsets = n;
+		policy->offsets = palloc(sizeof(policy->offsets[0]) * policy->num_allocated_offsets);
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/grouping_policy_hash.c:508]: We don't use repalloc because the data is per-batch and we don't care about the old values. I also added a pfree. The context is the ExecutorState where the vector_agg_exec is performed, as with most other allocations here.

File: tsl/src/nodes/vector_agg/function/functions.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply tsl/src/nodes/vector_agg/function/functions.c:61]: Yeah, so what I suggested is `num_rows`, which has clear semantics. Then you can have a "standard" for-loop  on i (starting at 0) and just get the row from start_row + i.

Not super-important, but the semantics are a bit more clear. Feel free to decide yourself what works best....

**Review by erimatnor — APPROVED** [review]

File: tsl/src/nodes/vector_agg/exec.c [hunk 18]
```diff
@@ -179,17 +179,51 @@ vector_agg_begin(CustomScanState *node, EState *estate, int eflags)
...
+			DecompressContext *dcontext = &decompress_state->decompress_context;
+			CompressionColumnDescription *desc =
+				&dcontext->compressed_chunk_columns[col->input_offset];
+			col->value_bytes = desc->value_bytes;
 		}
 	}
 
 	/*
-	 * Currently the only grouping policy we use is per-batch grouping.
+	 * Determine which grouping policy we are going to use.
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/nodes/vector_agg/exec.c]: I moved it all to plan time, although as we discussed today on call, it doesn't affect the costs yet.
